rfcs/0096-embedded-rust-interface-api.md

```diff
@@ -0,0 +1,274 @@
+
+This requires an additional build step in `build.rs` to reference the existing C files using `cc-rs`:
+
+```rust
+use cc;
+fn main() {
+    cc::Build::new()
+        .include("../codegen/host/include")
+        .include("../runtime/include")
+        .file("../codegen/host/src/default_lib0.c")
+        .file("../codegen/host/src/default_lib1.c")
+        .compile("mlf");
+}
+```
+
+This builds the existing C code and C interface for reference in Rust using the `ffi` module:
+
+```rust
+
+```
+
+## C Wrapper Structs
+By wrapping the C types in Rust structs, and exposing a `new` constructor it allows us to take a defined Rust array and translate it into a `void*`. Users of this interface only need to deal with pure Rust:
+
+```rust
+/// Input tensors for TVM module "rusty_coffee"
+#[repr(C)]
+pub struct Inputs {
+    input: *mut ::std::os::raw::c_void,
+}
+
+impl Inputs {
+    pub fn new <'a>(
+        input: &mut [u8; 100],
+    ) -> Self {
+        Self {
+            input: input.as_ptr() as *mut ::std::os::raw::c_void,
+        }
+    }
+}
+```
+
+For devices, we can only store the void pointer as it's passed without type information through the C runtime:
+```rust
+/// Device context pointers for TVM module "rusty_coffee"
+#[repr(C)]
+pub struct Devices {
+    ethos_u: *mut ::std::os::raw::c_void,
+}
+
+impl Devices {
+    pub fn new <'a>(
+        ethos_u: *mut ::std::os::raw::c_void,
+    ) -> Self {
+        Self {
+            ethos_u: ethos_u,
+        }
+    }
+}
+```
+
+## C Wrapper Entrypoint
+Similar to the above, using the Rust FFI we can create an entrypoint function which passes the void pointers directly into C FFI from the Rust structs with the appropriate `unsafe` block to prevent user facing code from having to be `unsafe`. The `run` wrapper also checks the return code of TVM and converts it into a [standard Rust `Result` object](https://doc.rust-lang.org/rust-by-example/error/result.html).
+
+```rust
+/// Entrypoint function for TVM module "rusty_coffee"
+/// # Arguments
+/// * `inputs` - Input tensors for the module
+/// * `outputs` - Output tensors for the module
+pub fn run(
+    inputs: &mut Inputs,
+    outputs: &mut Outputs,
+) -> Result<(), ()> {
+    unsafe {
+        let ret = tvmgen_rusty_coffee_run(
+            inputs,
+            outputs,
+        );
+        if ret == 0 {
+            Ok(())
+        } else {
+            Err(())
+        }
+    }
+}
+
+extern "C" {
+    pub fn tvmgen_rusty_coffee_run(
+        inputs: *mut Inputs,
+        outputs: *mut Outputs,
+    ) -> i32;
+}
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+This introduces a second embedded API alongside the C API, but fundamentally we want users to be able to create applications in the language that best suits their needs.
+
+Whilst it is an innovation project, this will be supported with best efforts but may be lag in features behind the C interface API.
```

Comment:
Not to block this RFC but just curious, who will be on the hook to maintain and ensure that the embedded rust API is properly tested? Right now for example the Rust TVM bindings aren't really owned by anyone and the Go bindings aren't even run in CI.

Reply:
Would it make sense to add a section to the RFC that outlines the testing and regression prevention strategy?

Reply:
The distinction here is features not robustness, as this lands in TVM I would expect it to be tested and regression tested in line with https://github.com/apache/tvm/blob/main/docs/contribute/code_review.rst#id11 ?